fdbclient/NativeAPI.actor.cpp

```diff
@@ -6839,6 +6839,255 @@ ACTOR Future<Void> snapCreate(Database cx, Standalone<StringRef> snapCmd, UID sn
+		    .detail("Version", version)
+		    .detail("Format", static_cast<int>(format));
+
+		state std::vector<std::vector<Future<CheckpointMetaData>>> alternatives;
+		state std::vector<Future<Void>> fs;
+		state int i = 0;
+		state int j = 0;
+
+		try {
+			state std::vector<std::pair<KeyRange, Reference<LocationInfo>>> locations =
+			    wait(getKeyRangeLocations(cx,
+			                              keys,
+			                              CLIENT_KNOBS->TOO_MANY,
+			                              Reverse::False,
+			                              &StorageServerInterface::checkpoint,
+			                              span.context,
+			                              Optional<UID>(),
+			                              UseProvisionalProxies::False));
+
+			alternatives.resize(locations.size());
+			for (i = 0; i < locations.size(); ++i) {
+				// For each shard, all storage servers are checked, only one is required.
+				for (j = 0; j < locations[i].second->size(); ++j) {
+					alternatives[i].push_back(locations[i].second->getInterface(j).checkpoint.getReply(
+					    GetCheckpointRequest(version, keys, format)));
+				}
+				TraceEvent("GetCheckpointShardBegin")
+				    .detail("Range", locations[i].first.toString())
+				    .detail("Version", version)
+				    .detail("StorageServers", locations[i].second->description());
+			}
+
+			for (i = 0; i < alternatives.size(); ++i) {
+				fs.push_back(waitForAny(alternatives[i]));
+			}
+
+			choose {
+				when(wait(cx->connectionFileChanged())) { cx->invalidateCache(keys); }
+				when(wait(waitForAll(fs))) { break; }
+				when(wait(delay(timeout))) {
+					TraceEvent("GetCheckpointTimeout").detail("Range", keys.toString()).detail("Version", version);
+					cx->invalidateCache(keys);
+				}
+			}
+		} catch (Error& e) {
+			TraceEvent("GetCheckpointError").detail("Range", keys.toString()).error(e, true);
+			if (e.code() == error_code_wrong_shard_server || e.code() == error_code_all_alternatives_failed ||
+			    e.code() == error_code_connection_failed || e.code() == error_code_broken_promise) {
+				cx->invalidateCache(keys);
+				wait(delay(CLIENT_KNOBS->WRONG_SHARD_SERVER_DELAY));
+			} else {
+				throw;
+			}
+		}
+	}
+
+	std::vector<CheckpointMetaData> res;
+	for (i = 0; i < alternatives.size(); ++i) {
+		for (j = 0; j < alternatives[i].size(); ++j) {
+			if (alternatives[i][j].isReady()) {
+				TraceEvent("GetCheckpointShardEnd").detail("Checkpoint", alternatives[i][j].get().toString());
+				res.push_back(alternatives[i][j].get());
+				break;
+			}
+		}
+		if (j >= alternatives[i].size()) {
+			throw internal_error();
+		}
+	}
+	return res;
+}
+
+// Fetch a single sst file from storage server. If the file is fetch successfully, it will be recorded via cFun.
+ACTOR static Future<Void> fetchCheckpointFile(Database cx,
+                                              std::shared_ptr<CheckpointMetaData> metaData,
+                                              int idx,
```

Comment:
This is basically building a file transfer protocol directly into FDB? I see several major issues with this approach:
- Allowing a client to stream arbitrary files directly through the native API seems like a major security hole, especially since we do not presently have any concept of authorization for storage server access. At minimum we need to provide a way to disable these endpoints on the SS before allowing this to merge. Snowflake would definitely run with this disabled, and I'd advocate for it to be disabled by default as a knob.
- This is going to be significantly more expensive than alternative methods of file transfer. File copying protocols like rsync have a lot more development and are going to be far more efficient. This will bottleneck on the FDB network thread client and SS side, and use up a non-trivial portion of those network threads.
- This is exposing rocksDB internals directly through the native API. This just seems like bad practice in general, as we'll have to maintain these structures in a backwards-compatible way through FDB versions AND ensure they remain up to date with RocksDB development. It overall is just very abstraction-breaking

Reply:
I just finished reading the rest of this PR.
What exactly is the purpose of this change?
If it's to speed up data movement, I'm not even convinced that this will help, and it may hurt.

Performance-wise:
On the source SS side, there is overhead to the checkpoint, and you will actually read MORE bytes from disk with this approach than you would doing a range read of that range from rocks, since you have to read and copy the entirety of each SST that overlaps the range. This will be especially problematic at the top levels of the LSM. You save the cpu of merging those SSTs on this side, but if you merged them you'd send even fewer bytes over the network with the materialized data than with the un-materialized data. You MAY make this up if the SST files are compressed.

Then, on the destination side, you get to bypass the rocks WAL, but you still have to de-compress and materialize the data, just in compaction instead of at transfer time. And, because of the over-read of these SST files, you have to over-compact to clean it up as well. And with compaction being the large majority of I/O for an LSM, I would not be surprised if that ended up being more expensive than just writing the data to the WAL.

tl;dr: if you are seeing data movement issues with rocks, I don't think this will help much, if at all, and may in fact make it worse. Likely what you are seeing then is just general data movement problems with FDB.

Reply:
That's a good point on the file transfer security issue, I will move this fetchCheckpoint to the server package.
Flow is chosen as the file transfer method since it has been well tested in FDB, so that it would be easier to find out potential issue/bugs in other changes. In the long-term, other protocols will be used if this hits the bottleneck.
This physical shard move will be deployed together with a sharded Rocks storage engine. This will allow us to avoid using the storage server read APIs when transferring data between storage servers.

Reply:
I think if this feature significantly helps data movement speed then the biggest takeaway is that FDB has far too much overhead in the Data Movement process and we should look for ways to reduce that.  Overheads that come to mind, in execution order, are
- Reading the range from IKVS
- Serializing the range result to a packet buffer
- FlowTransport for sending and receiving
- Deserializing the range result
- Writing _each_ KV pair, one at a time, to IKVS (fortunately the PTree is bypassed for data moves)

I think the last one is the largest overhead by far.  The Checkpoint feature avoids this, but I share Josh's concerns about moving an uncompacted slice of an LSM from one instance to another.  I don't think compacting first solves that either because then the range is being read, merged, rewritten, then read again to send to the remote side which will then write it again.  If data is being moved away due to high write load this will add to that load.

As an alternative, I think adding a `rangeReplace(RangeResult)` API to IKVS would be a good optimization, and some storage engines could do more optimization with this than others.  The fallback default implementation would be to clear and iterate/insert using existing IKVS methods.  

The next evolution would could be `readRangeNative(range)` and `replaceRangeNative(range, data)` where `data` is just a Type and an opaque blob only understood by like-minded storage engines.  

I think it is still valuable for the Storage Servers to execute data moves in increments of multi-megabyte chunks of data as it does now, but we should be able to improve the APIs for doing this where the increments are still small but have far less read / transfer / write overhead.

Reply:
And also, this PR and physical data move approach is not based on the assertion that it is superior to the current approach, it is move based on the new `sharded` data layout. Also an attempt to bring more options for FDB data move.

Reply:
Under each IKeyValueStore, there are multiple Rocks instances, each responsible for a distinct keyrange.